config/v1/types_dns.go

```diff
@@ -96,3 +103,44 @@ type DNSList struct {
 
 	Items []DNS `json:"items"`
 }
+
+// DNSPlatformType is a cloud infrastructure provider that supports
+// platform specific DNS configuration.
+// +kubebuilder:validation:Enum=AWS
+type DNSPlatformType PlatformType
+
+const (
+	// AWSDNSPlatformType represents Amazon Web Services.
+	AWSDNSPlatformType DNSPlatformType = "AWS"
+)
+
+// DNSPlatformSpec holds cloud-provider-specific configuration
+// for DNS administration.
+// +union
+// +kubebuilder:validation:XValidation:rule="has(self.platform) && self.platform == 'AWS' ?  has(self.aws) : !has(self.aws)",message="aws configuration is required when platform is AWS, and forbidden otherwise"
+type DNSPlatformSpec struct {
+	// type is the underlying infrastructure provider for the cluster.
+	// Allowed values: "AWS".
```

Comment:
@deads2k suggested allowed values be `AWS` and `""`, but that seems to contrast with the validation suggested [here](https://github.com/openshift/api/pull/1460/files#r1202637837), which is what I went with. 

The intent being: `dns.spec.platform` is optional, but if it is provided `dns.spec.platform.type` is required

Reply:
Empty needs to be allowed to handle upgrade cases.

Reply:
I have handled this in the code, but we couldn't figure out a test case (which you mentioned in slack). None of the fields in this config are mutable (although I'm not certain that is enforced). 

Was trying a test like this:

```yaml
  onUpdate:
  - name: Should allow empty platform on upgrade
    initial: |
      apiVersion: config.openshift.io/v1
      kind: DNS
      spec:
        zone:
          id: foo
    updated: |
      apiVersion: config.openshift.io/v1
      kind: DNS
      spec:
        zone:
          id: bar
    expected: |
      apiVersion: config.openshift.io/v1
      kind: DNS
      spec:
        zone:
          id: bar
        platform:
          type: ""
```

But that is failing 
>  [FAILED] the following fields were expected to match but did not:
  [(spec.platform/spec.platform) (spec.zone/spec.zone)]

We're not really sure how to write the upgrade test